model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java

```diff
@@ -537,40 +535,43 @@ public void removeLocalUserSessions(String realmId, boolean offline) {
 
         final AtomicInteger userSessionsSize = new AtomicInteger();
 
-        localCacheStoreIgnore
+        removeEntriesByRealm(realmId, localCacheStoreIgnore, userSessionsSize, localCache, localClientSessionCache);
```

Comment:
I am not following why we iterate over entries in `localCacheStoreIgnore` but remove from `localCache`. Since each KC node is doing it, wouldn't it be enough to iterate over `localCache` only?

Reply:
I suppose it is trying to do the following: 

* Get me all entries, but don't ask the remote store for the entries as it would be expensive
* If I then remote items, remove them als from the store which is remote cache

Reply:
> Get me all entries, but don't ask the remote store for the entries as it would be expensive

Yes that is correct, however, the event is processed by each Keycloak node and this means that each ID is processed also by nodes that do not own the entry which results in 3 removals of the same ID (in 3 KC node cluster). Then we use `localCache.removeAsync` which succeeds only on nodes that own the entry as we use the flag `CACHE_MODE_LOCAL` all other nodes do nothing.

I just realized the reason why we are doing it this way may be that we need to remove also client sessions which is making things a little complicated as client sessions may be owned by other nodes than user sessions and we also use `CACHE_MODE_LOCAL` for removal. I am thinking about a few things here:
1. What happens if some n1 owns `clientSession1` and n2 owns `userSession1`. Is it possible that n2 removes `userSession1`, without removing `clientSession1` before n1 iterates over `userSession1` which would mean `n1` never get to `userSession1` and hence never removes `clientSession1`?
2. I am not sure it makes sense to use `CACHE_MODE_LOCAL` for client sessions. Maybe we can use `SKIP_CACHE_STORE` and we can execute the removal only on primary owners for a user session.
3. Also, maybe we don't need to be afraid of keeping client sessions in there since they will be evicted eventually and they cannot be used without the corresponding user session. WDYT? 

> If I then remove items, remove them als from the store which is remote cache

If we talk only about this line, without the line 541, I don't see where we remove also from the remote cache. We are using `CACHE_MODE_LOCAL` for all removals which remove only from the current node if I am not mistaken.

Reply:
We discussed this in a meeting and we decided not to act on this comment at the moment as it would be just a performance optimization. The important part of this PR is to fix the possible issue when the remote store contains a different set of user sessions which is addressed by iterating also over the remote store. 

We can continue with the performance optimization with a lower priority later.